hibernate-core/src/main/java/org/hibernate/sql/results/jdbc/internal/JdbcValuesCacheHit.java

```diff
@@ -173,7 +173,10 @@ public Object getCurrentRowValue(int valueIndex) {
 			return null;
 		}
 		final Object row = cachedResults.get( position + offset );
-		if ( valueIndexesToCacheIndexes == null ) {
+		if ( row == null ) {
+			return null;
+		}
+		else if ( valueIndexesToCacheIndexes == null ) {
```

Comment:
I'm guessing the NPE was triggered by `row.getClass() != Object[].class`. Your change works, but I think we can use `instanceof` to be more concise:

```suggestion
		if ( valueIndexesToCacheIndexes == null || valueIndexesToCacheIndexes.length == 0 ) {
			return ( (Object[]) row )[valueIndex];
		}
		else if ( row instanceof Object[] ) {
			return ( (Object[]) row )[valueIndexesToCacheIndexes[valueIndex]];
		}
		else {
			assert valueIndexesToCacheIndexes[valueIndex] == 0;
			return row;
		}
```

Also, from looking at the code `valueIndexesToCacheIndexes` will never be null in the standard `JdbcValuesMapping` implementation - it will instead be an empty array, so it might be worth checking that as well.